Geometry/HGCalGeometry/src/HGCalGeometry.cc

```diff
@@ -315,38 +317,93 @@ HGCalGeometry::CornersVec HGCalGeometry::getCorners(const DetId& id) const {
+      xy = m_topology.dddConstants().locateCell(id.iLay,id.iSec1,id.iSec2,
+						id.iCell1,id.iCell2,true,false);
     }
-    float dx = ((m_det == DetId::HGCalHSc) ? 0.5*m_cellVec2[cellIndex].param()[11] :
-		m_cellVec[cellIndex].param()[1]);
-    float dz = ((m_det == DetId::HGCalHSc) ? m_cellVec2[cellIndex].param()[0] :
-		m_cellVec[cellIndex].param()[0]);
     static const int signx[] = {-1,-1,1,1,-1,-1,1,1};
     static const int signy[] = {-1,1,1,-1,-1,1,1,-1};
     static const int signz[] = {-1,-1,-1,-1,1,1,1,1};
-    for (unsigned int i = 0; i < ncorner; ++i) {
-      const HepGeom::Point3D<float> lcoord(xy.first+signx[i]*dx,
-					   xy.second+signy[i]*dx,signz[i]*dz);
-      co[i] = ((m_det == DetId::HGCalHSc) ? 
-	       (m_cellVec2[cellIndex].getPosition(lcoord)) :
-	       (m_cellVec[cellIndex].getPosition(lcoord)));
+    if (m_det == DetId::HGCalHSc) {
+      std::pair<double,double> rr = m_topology.dddConstants().cellSizeTrap(id.iType,id.iSec1);
+      float dx = 0.5*(rr.second-rr.first);
+      float dy = 0.5*(rr.second+rr.first)*m_cellVec2[cellIndex].param()[11];
+      float dz = m_cellVec2[cellIndex].param()[0];
+      for (unsigned int i = 0; i < ncorner; ++i) {
+	const HepGeom::Point3D<float> lcoord(signx[i]*dx,signy[i]*dy,signz[i]*dz);
+	co[i] = m_cellVec2[cellIndex].getPosition(lcoord);
+      }
+    } else {
+      float dx = m_cellVec[cellIndex].param()[1];
+      float dz = m_cellVec[cellIndex].param()[0];
+      for (unsigned int i = 0; i < ncorner; ++i) {
+	const HepGeom::Point3D<float> lcoord(xy.first+signx[i]*dx,
+					     xy.second+signy[i]*dx,signz[i]*dz);
+	co[i] = m_cellVec[cellIndex].getPosition(lcoord);
+      }
+    }
+  }
+  return co;
+}
+
+HGCalGeometry::CornersVec HGCalGeometry::getNewCorners(const DetId& detid) const {
+
+  unsigned int ncorner = (m_det == DetId::HGCalHSc) ? 5 : 7;
+  HGCalGeometry::CornersVec co (ncorner, GlobalPoint(0,0,0));
+  unsigned int cellIndex =  indexFor(detid);
+  if ((cellIndex <  m_cellVec.size() && m_det != DetId::HGCalHSc) ||
+      (cellIndex <  m_cellVec2.size() && m_det == DetId::HGCalHSc)) {
+    HGCalTopology::DecodedDetId id = m_topology.decode(detid);
+    int signz = (id.zSide >= 0) ? -1 : 1;
+    std::pair<float,float> xy;
+    if ((mode_ == HGCalGeometryMode::Hexagon) ||
+	(mode_ == HGCalGeometryMode::HexagonFull)) {
+      xy = m_topology.dddConstants().locateCellHex(id.iCell1,id.iSec1,true);
+    } else if (mode_ == HGCalGeometryMode::Trapezoid) {
+      xy = m_topology.dddConstants().locateCellTrap(id.iLay,id.iSec1,
+						    id.iCell1,true);
+    } else {
+      xy = m_topology.dddConstants().locateCell(id.iLay,id.iSec1,id.iSec2,
+						id.iCell1,id.iCell2,true,false);
+    }
+    if (m_det == DetId::HGCalHSc) {
+      std::pair<double,double> rr = m_topology.dddConstants().cellSizeTrap(id.iType,id.iSec1);
+      float dx = 0.5*(rr.second-rr.first);
+      float dy = 0.5*(rr.second+rr.first)*m_cellVec2[cellIndex].param()[11];
```

Comment:
Please don't use magic numbers like 5,7, and 11. Assign constants and use them instead.

Reply:
These numbers are standard in various types of calorimeter cells in CaloGeometry. I shall define the names in CaloGeometry package later on and transmit them to use cases in derived geometries.

Reply:
I created an issue for this task: https://github.com/cms-sw/cmssw/issues/25078